Add linting of source in markdown files

mkaz opened the pull request:
## Description

Adds [eslint-plugin-markdown](https://github.com/eslint/eslint-plugin-markdown) and the new command `lint-md` This allows for linting of code blocks that are marked js, javascript, or jsx, if a code block is not marked as such it should be skipped.

At the moment this uses the same base eslint rules, however we probably will want to adjust these rules for documentation. Most code blocks in documentation are snippets and incomplete chunks of code, so a fair amount of the current errors are due to `no-unused-vars` and `no-undef` rules. For example, a chunk of code might show how to set something, but not necessarily using what is set.
 
The first run through all docs:

```
✖ 2103 problems (2090 errors, 13 warnings)
```

## How has this been tested?

1. Test by applying this PR
2. Run: `npm install` to install new packages

To run across all documentation:
3a. Run `npm lint-md`

To run across a single file:
3b. $(npm bin)/wp-scripts lint-md docs/designers-developers/key-concepts.md


## Types of changes

1. Adds new script `lint-md` to wp-scripts
2. Adds eslint-plugin-markdown js package
3. Adds new command lint-md to main scripts

## Next Steps

We probably want to test out and adjust the rules to make sense for documentation.

Once we get it to a decent config, we can run it against the docs and fix up code that needs fixing and then include in the CI builds.

mkaz committed:
Add linting of markdown files

Adds eslint-plugin-markdown https://github.com/eslint/eslint-plugin-markdown
This allows for linting of code blocks marked js, javascript, or jsx if
a code block is not marked it is skipped.

mkaz committed:
Extension not needed, specified on command-line

ntwb commented:
Nice, and FYI, I’m working on the coding standards docs in https://github.com/WordPress/wpcs-docs repo (No pull request yet, working on it locally) and will be creating a new package here `@wordpress/remark-config` for a _shared config_ for Remark Lint to ensure markdown is compatible with the w.org handbooks

ntwb commented:
Further, I’ll look at what stylelint options are available to support listing CSS & SCSS in markdown files too

gziolo commented on package.json:
```diff
@@ -189,6 +189,7 @@
 		"lint-css": "wp-scripts lint-style '**/*.scss'",
 		"lint-css:fix": "npm run lint-css -- --fix",
 		"lint-types": "tsc",
+		"lint-md": "wp-scripts lint-md '**/*.md'",
```
I think 
```suggestion
		"lint-md": "wp-scripts lint-md",
```
should be enough since the code provides defaults.

gziolo commented on packages/scripts/scripts/lint-md.js:
```diff
@@ -0,0 +1,54 @@
+const defaultFilesArgs = hasFileArgInCLI() ? [] : [ '**/*.md' ];
+
+// See: https://eslint.org/docs/user-guide/configuring#using-configuration-files-1.
+const hasLintConfig = hasArgInCLI( '-c' ) ||
```
It probably could be extracted to a helper function to avoid code duplication, but more importantly, it will help to prevent diverging implementations. 

See:
https://github.com/WordPress/gutenberg/blob/50f471e0596a36d4b6a2df2b9cae382b85dde30b/packages/scripts/utils/config.js#L13-L28

mkaz replied:
This part is no longer needed, because I changed things around, creating a new config just for markdown to allow for overriding some of the rules.

ntwb replied:
One of these days if no one beats me to it is to swap all this out with [cosmiconfig(https://www.npmjs.com/package/cosmiconfig)

gziolo reviewed (approved):
This is great, thanks for opening it.

gziolo commented:
I think this PR addresses #12426, well at least the part for linting markdown.

> At the moment this uses the same base eslint rules, however we probably will want to adjust these rules for documentation. Most code blocks in documentation are snippets and incomplete chunks of code, so a fair amount of the current errors are due to `no-unused-vars` and `no-undef` rules. For example, a chunk of code might show how to set something, but not necessarily using what is set.

It might end up with having two ESLint configs for `wp-scripts`, since we will need to disable some rules. I would say, let's start simple and disable the noisiest errors and evaluate later the next steps. I'm thinking in terms of a follow-up PR, to be clear.

mkaz committed:
Update package.json

Co-Authored-By: Grzegorz (Greg) Ziółkowski <grzegorz@gziolo.pl>

mkaz committed:
Create and use new eslint config for markdown

- Creates a new eslintrc-md.js config for use for markdown docs

- The lint-md command uses this config unless another is specified

- The config allows for undef and unused vars due to docs contain
snippets of code that may be incomplete

mkaz committed:
Since using separate config, no change needed

mkaz committed:
Dont need markdown in main, its in wp-scripts
